object_intern.go

```diff
@@ -198,80 +170,43 @@ func (oi *ObjectIntern) AddOrGetCompressed(obj []byte) (uintptr, error) {
 }
 
 // AddOrGetString finds or adds an object and then returns a string with its Data pointer set to the newly interned object and nil.
-// This method assumes that compression is turned off.
+// This method takes a []byte of the object, and two bools. If compressed
+// is set to true then we assume compression is being used. If safe is set to true
+// then this method will create a copy of the []byte before performing any operations
+// that might modify the backing array.
 // On failure it returns an empty string and an error
 //
 // If the object is found in the store its reference count is increased by 1.
 // If the object is added to the store its reference count is set to 1.
-func (oi *ObjectIntern) AddOrGetString(obj []byte) (string, error) {
+func (oi *ObjectIntern) AddOrGetString(obj []byte, compressed, safe bool) (string, error) {
 	var addr gos.ObjAddr
 	var ok bool
 	var err error
-	var objString string
-
-	// create a copy so we don't modify the original, especially because compression is turned off
-	// when compression is turned on using SHOCO it will return a new slice, which is not the case here
-	objComp := make([]byte, len(obj))
-	copy(objComp, obj)
-	objString = string(obj)
-
-	// acquire lock
-	oi.Lock()
-
-	// try to find the object in the index
-	addr, ok = oi.ObjIndex[objString]
-	if ok {
-		// increment reference count by 1
-		(*(*uint32)(unsafe.Pointer(addr + uintptr(len(objComp)))))++
-		tmpString := objString
-		StringHeader := (*reflect.StringHeader)(unsafe.Pointer(&tmpString))
-		StringHeader.Data = addr
-		oi.Unlock()
-		return tmpString, nil
-	}
-
-	// The object is not in the index therefore it is not in the store.
-	// We need to set its initial reference count to 1 before adding it
-	objComp = append(objComp, []byte{0x1, 0x0, 0x0, 0x0}...)
-	addr, err = oi.Store.Add(objComp)
-	if err != nil {
-		oi.Unlock()
-		return "", err
+	var objComp []byte
+
+	// is the data supposed to be compressed?
+	if compressed {
+		// even so, is compression actually turned on, and do we need to be safe?
+		if oi.conf.Compression == None && safe {
+			// create a copy
+			objComp = make([]byte, len(obj))
+			copy(objComp, obj)
+		} else {
+			// if compression is turned on, this will return a new []byte
+			objComp = oi.compress(obj)
```

Comment:
the way how strings are stored when compression is enabled seem strange to me, but that might just be my jet lag and i don't get it:
in the index they're still stored in un-compressed form, instead of using the compressed string object which is backed by the object store in the index as well. doesn't that mean that everything will get stored once compressed and once uncompressed? what's the point of compressing it then if the interning needs to keep an uncompressed copy for the index.

Reply:
I have made a few changes which I will push soon. the index will store compressed or uncompressed versions of the data (depending on if compression is turned on). It will always return a decompressed version of the data if a string or []byte is requested though.